lib/libzpool/uzfs_mtree.c

```diff
@@ -0,0 +1,251 @@
+			umem_free(b_entry, sizeof (*b_entry));
+			goto find;
+		}
+
+		if (b_end < (new_offset + new_len)) {
+			new_len += (new_offset - b_entry->offset);
+			new_offset = b_entry->offset;
+			avl_remove(tree, b_entry);
+			umem_free(b_entry, sizeof (*b_entry));
+			goto find;
+		}
+
+		if (b_end >= (new_offset + new_len))
+			return;
+	}
+
+after:
+	a_entry = avl_nearest(tree, where, AVL_AFTER);
+
+	if (a_entry) {
+		a_end = (a_entry->offset + a_entry->len);
+		if ((new_offset + new_len) < a_entry->offset)
+			goto doadd;
+
+		if ((new_offset + new_len) == a_entry->offset) {
+			new_len += a_entry->len;
+			avl_remove(tree, a_entry);
+			umem_free(a_entry, sizeof (*a_entry));
+			goto find;
+		}
+
+		if ((new_offset + new_len) <= (a_end)) {
+			new_len = (a_entry->len) +
+			    (a_entry->offset - new_offset);
+			avl_remove(tree, a_entry);
+			umem_free(a_entry, sizeof (*a_entry));
+			goto find;
+		}
+
+		if ((new_offset + new_len) > (a_end)) {
+			avl_remove(tree, a_entry);
+			umem_free(a_entry, sizeof (*a_entry));
+			goto find;
+		}
+	}
+
+doadd:
+	new_node = umem_alloc(sizeof (uzfs_zvol_blk_phy_t), UMEM_NOFAIL);
+	new_node->offset = new_offset;
+	new_node->len = new_len;
+	avl_insert(tree, new_node, where);
+}
+
+void
+dump_mblktree(avl_tree_t *tree)
+{
+	uzfs_zvol_blk_phy_t *blk;
+
+	for (blk = avl_first(tree); blk; blk = AVL_NEXT(tree, blk)) {
+		printf("offset:%lu, length:%lu\n", blk->offset, blk->len);
+	}
+}
+
+int
+uzfs_changed_block_cb(spa_t *spa, zilog_t *zillog, const blkptr_t *bp,
+    const zbookmark_phys_t *zb, const dnode_phys_t *dnp, void *arg)
+{
+	uint64_t blksz;
+	struct diff_txg_blk *diff_blk_info = (struct diff_txg_blk *)arg;
+
+	if ((bp == NULL) || (BP_IS_HOLE(bp)) || (zb->zb_object != ZVOL_OBJ) ||
+	    (zb->zb_level != 0))
+		return (0);
+
+	if (bp->blk_birth > diff_blk_info->end_txg ||
+	    bp->blk_birth < diff_blk_info->start_txg)
+		return (0);
+
+	blksz = BP_GET_LSIZE(bp);
+	add_to_mblktree(diff_blk_info->tree, zb->zb_blkid * blksz, blksz);
+	return (0);
+}
+
+static int
+zvol_blk_off_cmpr(const void *arg1, const void *arg2)
+{
+	uzfs_zvol_blk_phy_t *node1 = (uzfs_zvol_blk_phy_t *)arg1;
+	uzfs_zvol_blk_phy_t *node2 = (uzfs_zvol_blk_phy_t *)arg2;
+
+	return (AVL_CMP(node1->offset, node2->offset));
+}
+
+
+int
+uzfs_txg_block_diff(zvol_state_t *zv, uint64_t start_txg, uint64_t end_txg,
+    avl_tree_t **tree)
+{
+	int error;
+	char snapname[ZFS_MAX_DATASET_NAME_LEN];
+	struct diff_txg_blk diff_blk;
+	hrtime_t now;
+	dsl_pool_t *dp;
+	dsl_dataset_t *ds_snap;
+
+	now = gethrtime();
+	snprintf(snapname, sizeof (snapname), "%s%llu", TXG_DIFF_SNAPNAME, now);
+
+	error = dmu_objset_snapshot_one(zv->zv_name, snapname);
+	if (error) {
+		printf("failed to create snapshot for %s\n", zv->zv_name);
+		return (error);
+	}
+
+	memset(snapname, 0, sizeof (snapname));
+	snprintf(snapname, sizeof (snapname), "%s@%s%llu", zv->zv_name,
+	    TXG_DIFF_SNAPNAME, now);
+
+	error = dsl_pool_hold(snapname, FTAG, &dp);
```

Comment:
why don't we use dmu_objset_own rather than dsl_pool_hold & dsl_dataset_hold?

Reply:
here, we are traversing snapshot only, so we can avoid dmu_objset_own. Regarding snapshot destruction while we are traversing snapshot, we can use dsl_dataset_long_hold.